rudder-lang/tools/generate_lib

```diff
@@ -1,139 +1,539 @@
+
+import re
+import json
+import os.path
+import sys
+import os
+import codecs
+try:
+    from ConfigParser import ConfigParser
+except BaseException:
+    from configparser import ConfigParser
+from pprint import pprint
+
+tags = [
+    "name",
+    "description",
+    "parameter",
+    "method",
+    "documentation",
+    "class_prefix",
+    "class_parameter",
+    "class_parameter_id",
+    "deprecated",
+    "agent_requirements",
+    "parameter_constraint",
+    "parameter_type",
+    "action",
+    "rename"]
+optionnal_tags = [
+    "deprecated",
+    "documentation",
+    "parameter_constraint",
+    "parameter_type",
+    "agent_requirements",
+    "action",
+    "rename"]
+multiline_tags = ["description", "documentation", "deprecated"]
+
+
+class NcfError(Exception):
+    def __init__(self, message, details="", cause=None):
+        self.message = message
+        self.details = details
+        # try to get details from inner cause
+        try:
+            # Will not add to details if cause is None or message is None
+            self.details += " caused by : " + cause.message
+            # Will not add to details if details is None
+            self.details += "\n" + cause.details
+        except BaseException:
+            # We got an error while extending error details, just ignore it and
+            # keep current value
+            pass
+
+    def __str__(self):
+        return repr(self.message)
+
+
+def format_errors(error_list):
+    formated_errors = []
+    for error in error_list:
+        sys.stderr.write("ERROR: " + error.message + "\n")
+        sys.stderr.write(error.details + "\n")
+        formated_errors.append(
+            {"message": error.message, "details": error.details})
+    sys.stderr.flush()
+    return formated_errors
+
+
+def get_methods_filenames(parent_dir):
+    filenames = []
+    filenames_add = filenames.append
+    for root, dirs, files in os.walk(parent_dir):
+        for dir in dirs:
+            filenames = filenames + \
+                get_methods_filenames(os.path.join(parent_dir, dir))
+
+        for file in files:
+            if file.endswith(".cf") and not file.startswith("_"):
+                filenames.append(os.path.join(root, file))
+    return filenames
+
+##############
+# data parsing
+##############
+
+# This function is adapted from the ncf lib
+
+
+def parse_method_metadata(content):
+    res = {}
+    warnings = []
+    parameters = {}
+    param_names = set()
+    param_constraints = {}
+    param_types = {}
+    default_constraint = {}
+
+    multiline = False
+    previous_tag = None
+    match_line = ""
+    bundle_name = "unknown"
+
+    for line in content.splitlines():
+        # line should already be unicode
+        # unicodeLine = unicode(line,"UTF-8") #line.decode('unicode-escape')
+
+        # Parse metadata tag line
+        match = re.match(
+            r"^\s*#\s*@(\w+)\s*(([a-zA-Z0-9_]+)?\s+(.*?)|.*?)\s*$",
```

Comment:
named parameters (if possible) would make it easier to understand

Reply:
i have to admit I just took the regexes from the ncf library

Reply:
but I can try to improve it indeed.